onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2832,101 @@ ONNX_OPERATOR_SET_SCHEMA(
+              AttributeProto::INT,
+              OPTIONAL_VALUE)
+        .Input(0,
+               "X",
+               "The input tensor representing the token embeddings. 3D tensor with shape (batch_size, sequence_length, head_size), head_size is supposed to be even",
+               "T")
+        .Input(1,
+               "position_ids",
+               "The position indices for the tokens. 1D tensor with shape (1) or 2D tensor with shape (batch_size, sequence_length)",
+               "M")
+        .Input(2,
+               "cos_cache",
+               "The cosine values for the rotation. 2D tensor with shape (max_sequence_length, head_size / 2)",
+               "T")
+        .Input(3,
+               "sin_cache",
+               "The sine values for the rotation. 2D tensor with shape (max_sequence_length, head_size / 2)",
+               "T")
+        .Output(0,
+                "Y",
+                "tensor with same shape as input.",
+                "T")
+        .TypeConstraint("T", {"tensor(float)", "tensor(float16)", "tensor(bfloat16)"}, "Constrain input and output types to float tensors.")
+        .TypeConstraint("M", {"tensor(int64)"}, "Constrain input and output types to integer tensors")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+        })
+        .SetContextDependentFunctionBodyBuilder(
+            [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+
+              auto mktensor = [](int64_t val) -> ONNX_NAMESPACE::TensorProto {
+                auto tp = ONNX_NAMESPACE::ToTensor(std::vector<int64_t>{val});
+                tp.add_dims(1);
+                return tp;
+              };
+
+              FunctionBuilder builder(functionProto);
+              builder.Add("CosCacheGather = Gather(cos_cache, position_ids)") // shape of cos_matrix: [b, seq_len, dim]
+                .Add("CosCacheUnsqueezed = Unsqueeze(CosCacheGather)") // shape of cos_matrix: [b, 1, seq_len, dim]
+                .Add("CosCacheGather = Gather(CosCacheSqueezed, position_ids)") // shape of sin_matrix: [b, seq_len, dim]
+                .Add("SinCacheUnsqueezed = Unsqueeze(sin_cache, UnsqueezeDims)"); // shape of sin_matrix: [b, 1, seq_len, dim]
+
+              builder.Add("Two1D = Constant()", "value", mktensor(2)) // [2] : 1D tensor
+                  .Add("Shape = Shape (X)") // shape of input tensor: 1D tensor
+                  .Add("RotateEmbedDim = Gather(Shape, Two1D)") // 1D tensor
+                  .Add("RotateEmbedDimHalf = Div(RotateEmbedDim, Two1D)")
+                  .Add("ExpandShape = Constant()", "value", mktensor(2))
+                  .Add("SplitTensor = Expand(RotateEmbedDimHalf, ExpandShape)")
+                  .Add("InputFirstHalf, InputSecondHalf = Split <axis = -1> (X, SplitTensor)")
```

Comment:
Specify `num_outputs = 2` attribute value for the split. I don't think you need specify `SplitTensor`. Much simpler, many of preceding statements/nodes are not needed.